src/header-generator.js

```diff
@@ -0,0 +1,271 @@
+     */
+    getHeaders(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        const headerOptions = { ...this.defaultOptions, ...options };
+
+        // Set up defaults
+        if (!headerOptions.locales) {
+            headerOptions.locales = ['en-US'];
+        }
+        if (!headerOptions.httpVersion) {
+            headerOptions.httpVersion = '2';
+        }
+        if (!headerOptions.browsers) {
+            headerOptions.browsers = [
+                { name: 'chrome' },
+                { name: 'firefox' },
+                { name: 'safari' },
+            ];
+        }
+        if (!headerOptions.operatingSystems) {
+            headerOptions.operatingSystems = [
+                'windows',
+                'macos',
+                'linux',
+                'android',
+                'ios',
+            ];
+        }
+
+        headerOptions.browsers = headerOptions.browsers.map((browserObject) => {
+            if (!browserObject.httpVersion) {
+                browserObject.httpVersion = headerOptions.httpVersion;
+            }
+            return browserObject;
+        });
+
+        const possibleAttributeValues = {};
+
+        // Find known browsers compatible with the input
+        const browserHttpOptions = [];
+        for (const browser of headerOptions.browsers) {
+            for (const browserOption of this.uniqueBrowsers) {
+                if (browser.name === browserOption.name) {
+                    if ((!browser.minVersion || browserVersionIsLesserOrEquals([browser.minVersion], browserOption.version))
+                        && (!browser.maxVersion || browserVersionIsLesserOrEquals(browserOption.version, [browser.maxVersion]))
+                        && browser.httpVersion === browserOption.httpVersion) {
+                        browserHttpOptions.push(browserOption.completeString);
+                    }
+                }
+            }
+        }
+
+        possibleAttributeValues[browserHttpNodeName] = browserHttpOptions;
+
+        possibleAttributeValues[operatingSystemNodeName] = headerOptions.operatingSystems;
+
+        if (headerOptions.devices) {
+            possibleAttributeValues[deviceNodeName] = headerOptions.devices;
+        }
+
+        // Generate a sample of input attributes consistent with the data used to create the definition files if possible.
+        const inputSample = this.inputGeneratorNetwork.generateSampleWheneverPossible(possibleAttributeValues);
+
+        if (!inputSample) {
+            throw new Error('No headers based on this input can be generated. Please relax or change some of the requirements you specified.');
+        }
+
+        // Generate the actual headers
+        const generatedSample = this.headerGeneratorNetwork.generateSample(inputSample);
+
+        // Manually fill the accept-language header with random ordering of the locales from input
```

Comment:
Maybe we can put the accept-language to a separate function?

Reply:
Going from that logic, this method should be splintered to 4-5 functions. I prefer it this way, since a reader can read it all in uninterrupted progression without jumping around to different function, but I get that somebody might dislike it. Do you think is should be changed?

Reply:
That makes sense. I enjoyed reading it, but it was here when the chapter got kind of long. But this is definitely a matter of opinion. I think we can keep it this way.